erpnext/regional/doctype/import_supplier_invoice/import_supplier_invoice.py

```diff
@@ -0,0 +1,386 @@
+
+		if pi_count == file_count:
+			self.status = "File Import Completed"
+			self.publish("File Import", _("XML Files Processed"), 2, 3)
+		else:
+			self.status = "Partially Completed - Check Error Log"
+			self.publish("File Import", _("XML Files Processed"), 2, 3)
+
+		self.save()
+		self.publish("File Import", _("XML Files Processed"), 3, 3)
+
+	def process_file_data(self):
+		self.status = "Processing File Data"
+		self.save()
+		frappe.enqueue_doc(self.doctype, self.name, "import_xml_data", queue="long", timeout=3600)
+
+	def publish(self, title, message, count, total):
+		frappe.publish_realtime("import_invoice_update", {"title": title, "message": message, "count": count, "total": total})
+
+def create_supplier(**args):
+
+	args = frappe._dict(args)
+	existing_supplier_name = frappe.db.get_value("Supplier",
+				filters={"tax_id": args.tax_id}, fieldname="name")
+	if existing_supplier_name:
+		pass
+	else:
+		existing_supplier_name = frappe.db.get_value("Supplier",
+				filters={"name": args.supplier}, fieldname="name")
+
+	if existing_supplier_name:
+		filters = [
+				["Dynamic Link", "link_doctype", "=", "Supplier"],
+				["Dynamic Link", "link_name", "=", args.existing_supplier_name],
+				["Dynamic Link", "parenttype", "=", "Contact"]
+			]
+
+		existing_contacts = frappe.get_list("Contact", filters)
+
+		if existing_contacts:
+			pass
+		else:
+			new_contact = frappe.new_doc("Contact")
+			new_contact.first_name = args.supplier
+			new_contact.append('links', {
+				"link_doctype": "Supplier",
+				"link_name": existing_supplier_name
+			})
+			new_contact.insert()
+
+		return existing_supplier_name
+	else:
+		
+		new_supplier = frappe.new_doc("Supplier")
+		new_supplier.supplier_name = args.supplier
+		new_supplier.supplier_group = args.supplier_group
+		new_supplier.tax_id = args.tax_id
+		new_supplier.fiscal_code = args.fiscal_code
+		new_supplier.fiscal_regime = args.fiscal_regime
+		new_supplier.save()
+
+		new_contact = frappe.new_doc("Contact")
+		new_contact.first_name = args.supplier
+		new_contact.append('links', {
+			"link_doctype": "Supplier",
+			"link_name": new_supplier.name
+		})
+
+		new_contact.insert()
+
+		return new_supplier.name
+
+def create_address(**args):
+
+	args = frappe._dict(args)
+	filters = [
+			["Dynamic Link", "link_doctype", "=", "Supplier"],
+			["Dynamic Link", "link_name", "=", args.supplier_name],
+			["Dynamic Link", "parenttype", "=", "Address"]
+		]
+
+	existing_address = frappe.get_list("Address", filters)
+
+	if args.address_line1:
+		make_address = frappe.new_doc("Address")
+		make_address.address_line1 = args.address_line1
+
+		if args.city:
+			make_address.city = args.city
+		else:
+			make_address.city = "Not Provided"
+
+		if args.province:
+			make_address.state_code = args.province
+
+		if args.pincode:
+			make_address.pincode = args.pincode
+
+		if args.country:
+			make_address.country = args.country
```

Comment:
```
for field in ["city", "province", "pincode", "country"]:
    if args.get(field):
        new_address_doc.set(field, args.get(field))
```

Reply:
City cannot be included as it is mandatory and needs a default value if not provided, other values can follow this pattern